In [1]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression

In [2]:
import pandas as pd

In [3]:
#Imports NYC DF - DC/EDA
NYCDF = pd.read_csv('NYCDF.csv')

In [4]:
#Imports Boston DF - DC/EDA
BostonDF = pd.read_csv('BostonDF.csv')

In [5]:
#Combines Boston and NYC DF - PPM
DFBoth = pd.concat([NYCDF, BostonDF], axis=0)

In [6]:
#Confirms DF created - PPM
DFBoth

Title  \
0      Monthly Discussion Thread - Month of July, 2019   
1               This is what McDonald’s in NYC be like   
2            I hate how this city deals with its trash   
3    do you ever feel like you're not taking advant...   
4    Delancey Street in Lower Manhattan, c. 1961. -...   
5    An Open Letter to The People of New York City....   
6                                  missing teen in nyc   
7    $399 for a Festival in Central Park? The Great...   
8    Jeffrey Epstein Taught at Dalton. His Behavior...   
9    Pro-immigrant NYC residents vastly outnumber R...   
10   Man busted for driving in HOV lane with dummy ...   
11                                      Looking north.   
12                              to “protect and serve”   
13   17-year-old sign spotted at the 95th St R plat...   
14   What’s the thing with chess players in open in...   
15                                        Good evening   
16   Michael Seidenberg, NYC's Speakeasy Bookseller...   
17                       7th Precinct: Parking Experts   
18   Taking my Road Test in Fresh Meadows next mont...   
19                                           Lost item   
20   'Cash Cab' cruising back to New York City this...   
21   Can you add a second kitchen to a house and if...   
22   Cuomo Calls on MTA to Address Homelessness in ...   
23   Get Your Free Bike Helmet at Sara D. Roosevelt...   
24   [PDF] MTA Inspector General Releases Reports O...   
25   They Took the Bench and the Rain Cover from th...   
26   MTA crackdown: Express bus rider claims he was...   
27                  anyone midtown have a floppy disc?   
28   The elevators on the World Trade Center tour s...   
29   Transit union boss whines about coverage of ch...   
..                                                 ...   
965  Did you see a bike/pedestrian accident near th...   
966                              101 Seaport Boulevard   
967  For real, why don't more people call Revere Be...   
968               Best public beaches south of Boston?   
969  Longtime radio personality Gerry Callahan out ...   
970  On the Hook: Cops Accuse Tow Truck Operator of...   
971  20min north of Boston in my apartment building...   
972                 Fight on Red Line Train at Kendall   
973                     My view from my meeting today.   
974                        Anybody recognize this bar?   
975                                   Castle Island :)   
976  Skyline &amp; Charles River looking southwest ...   
977  Mass. OKs Air Permit For Controversial Weymout...   
978  An Augmented Reality Experience On The Greenwa...   
979                            Long Exposure of Boston   
980  Boston Globe Accused of ‘Willie Horton’-Style ...   
981           I’m mad as hell, but I won’t ditch the T   
982                 Confessions of a Red Line Defector   
983                                    Red line delays   
984  Sorry if this is the wrong place, if so please...   
985  Gentle reminder that the first wave of raids s...   
986  As threat of raids loom, immigrants in Boston ...   
987                      Conveyor belt sushi in Boston   
988  [Question] I have balcony tickets to a show at...   
989  Bill Allowing ‘X’ Gender On NH Licenses Become...   
990  New Hampshire woman told she can't fly 'Trump ...   
991                     Looking for a powerlifting gym   
992  Found Nick’s keys on the ground in front of Na...   
993                                        scoot scoot   
994         Coffee Shop-Style Cafe that serves Alcohol   

                                              Comments Subreddit  
0    Hello! This thread is for discussions, questio...       nyc  
1                                                  NaN       nyc  
2                                                  NaN       nyc  
3    I've lived here nearly 3 years and I find myse...       nyc  
4                                                  NaN       nyc  
5                                                 

In [7]:
#Gives each Subreddit a value of 1 or 0 depending on BOS or NYC - PPM
DFBoth['Subreddit'] = DFBoth['Subreddit'].map({'boston': 0, 'nyc': 1})
DFBoth.head()

Title  \
0    Monthly Discussion Thread - Month of July, 2019   
1             This is what McDonald’s in NYC be like   
2          I hate how this city deals with its trash   
3  do you ever feel like you're not taking advant...   
4  Delancey Street in Lower Manhattan, c. 1961. -...   

                                            Comments  Subreddit  
0  Hello! This thread is for discussions, questio...          1  
1                                                NaN          1  
2                                                NaN          1  
3  I've lived here nearly 3 years and I find myse...          1  
4                                                NaN          1

In [8]:
#Gets value counts of dataframes - PPM
DFBoth['Subreddit'].value_counts(normalize=True)

0    0.504308
1    0.495692
Name: Subreddit, dtype: float64

In [13]:
#Gets number of Boston vs NYC subreddits - PPM
(DFBoth['Subreddit'] == 0).sum(), (DFBoth['Subreddit'] == 1).sum()

(995, 978)

In [14]:
def basic_model(title_obs):
    title = str(title_obs)
    if 'nyc' in title.lower():
        return 1
    else:
        return 0

In [31]:
#Basic Model to build off of - PPM
basic_model("I love pizza in NYC!"), basic_model("I love pizza in boston!")

(1, 0)

In [17]:
#Set predictions - PPM
y_pred = [basic_model(title) for title in list(DFBoth['Title'])]

In [19]:
#Sets actual - PPM
y_actual = list(DFBoth['Subreddit'])
y_actual[:5]

[1, 1, 1, 1, 1]

In [26]:
#Sets total correct - PPM
total_correct = sum([1 if _pred == _actual else 0 
                    for _pred, _actual in zip(y_pred, y_actual)]
                     )
total_correct

1152

In [29]:
#Sets total observations - PPM
total_observations = len(y_actual)
total_observations

1973

In [30]:
#Gets accuracy - PPM
accuracy = total_correct / total_observations
accuracy

0.5838824125696909

So just looking for word 'nyc', we have 58% accuracy - meaning we get 1152 correct out of 1973 observation correct. As stated earlier:

you need to use CountVectorizer on the `Title` column for each row

That should make a df with thousands of columns, one for each word

Then you'll do `LogisticRegression` on that many column df, with those cols as X, and `Subreddit` column as your y

In [9]:
dfTFIDF  = pd.DataFrame(tvec.fit_transform(texts).toarray(),
                   columns=tvec.get_feature_names())
df.head()

NameError: name 'tvec' is not defined